### ISSUE TITLE
Section id with characters outside of Windows-1252 encoding causes warning

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor-pdf#308]
If a section id (auto-generated or otherwise) has characters outside of Windows-1252 encoding, a warning is issued. This doesn't seem to break xrefs, but it probably means we need to hex encode them so that they are guaranteed to work in all viewers.

Example:

```
= Document Title

== Проверка

content

<<<

See <<_Проверка>>.
```

**Comment by mojavelinux** [comment]
The solution to this problem is to "hexify" the internal anchors for xrefs. What I can't decide is whether we should do this globally or only if non-ascii characters are detected. I'm leaning towards doing it for all values so it's consistent.

(It would be nice if Prawn did this for us).

**Comment by flarcher** [comment]
I have a similar (same?) issue with a special french character in a section title. But instead of printing a warning, it simply stops the rendering and the PDF gets filled with the _asciidoc_ text (instead of the corresponding PDF content) starting at the point of the section start (where the special character lies).

I am using `org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.9` (same version as @orloffm in #69) from _Linux Ubuntu_.

**Comment by mojavelinux** [comment]
Can you please provide the sample file. It's best if you are able to serve it over a webserver (your own or Dropbox)? Copy and pasting the content into a web form usually changes the encoding of the text...and hence we are back to square one trying to debug it. Even better is if you can zip it up so as to protect the file from being modified.

**Comment by flarcher** [comment]
Thanks for that fast answer !

Here is the input: 
```
= Hello
blabla
== hehê
hello
image::data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==[]
```
The command is:
```
asciidoctor -B /home/fabrice/tmp -S SAFE -o /home/fabrice/tmp/test.pdf -b pdf -a docdate=2015-09-11 -a data-uri! -a allow-uri-read -a doctime=09:45:04 CEST -a icons=font -
```
And here is the output : https://drive.google.com/file/d/0BwNCc1A1YY17UVlvbU9WclVabUU/view?usp=sharing

**Comment by flarcher** [comment]
The problem that I am facing seems unrelated to the special character afterall :confused: (It also occurs with `== hehe` instead of `== hehê`).

I discovered that I should insert empty lines between each statement. And it seems that it causes the problem that I noticed.

**Comment by mojavelinux** [comment]
> Is that a known issue ? (Should I move my comments into a new issue ?)

No need. We can continue working in this issue. I'll look into the input you provided and try to explain what's happening.

**Comment by mojavelinux** [comment]
> I discovered that I should insert empty lines between each statement.

Each block should be separate by a blank line. This is just the way the AsciiDoc syntax is designed (blank lines are often significant).

In your case, the line `blabla` is assumed to be the author and the line `== hehê` is considered to be a revision line. That's why you are seeing the result that you shared.

It's especially important that there is a blank line between the document header and the beginning of the content (just like HTTP). See http://asciidoctor.org/docs/user-manual/#doc-header.

**Comment by mojavelinux** [comment]
So you would write your sample as:

```
= Hello

blabla

== hehê

hello

image::/home/fabrice/tmp/test-image-1.png[]
```

**Comment by SandMann1** [comment]
I don't know if this is connected to this issue or if it shoud be a new issue. I get the same warning "..not be fully converted to the Windows-1252 character set" when the character is within a kbd:[] 

Example: 
kbd:[↑] results in a ¬ in the pdf. . 

When the upward arrow glyph is passed "on its own" - no problem. Using aplha 11.

**Comment by mojavelinux** [comment]
Are you using your own theme?

If you are getting this warning, it's very likely that you are using the built-in font names in your PDF theme. The built-in fonts only support characters in the Windows-1252 character set (as explained here: https://github.com/asciidoctor/asciidoctor-pdf/blob/master/docs/theming-guide.adoc#built-in-afm-fonts). You must use your own font (or the fonts provided in the default theme) if you want to output characters beyond this set (and get proper UTF-8 support).

**Comment by mojavelinux** [comment]
So this issue is different. This issue is about how ids are stored internally, which triggers a warning even if the character output fine.

**Comment by SandMann1** [comment]
Ok. I am using a custum theme. I also have my own fontsdir and fonts, but I can see something is wrong there. Tryed the dafualt theme and everything is ok there. Thanks for quick response as always.

**Comment by mojavelinux** [comment]
Super. If you need help with the theme, don't hesitate to ask. I know dealing with custom fonts could still be a lot smoother, so I'm going to keep working on making it easier and the information clearer.

**Comment by SandMann1** [comment]
Related question, my custum theme is attached here:
[my-theme.zip](https://github.com/asciidoctor/asciidoctor-pdf/files/236580/my-theme.zip)

In my language we use special characters as Ææ Øø and Åå. My custom fonts supports these characters and they are printet fine in the resulting pdf. My problem appears when my end users wants to copy and paste text from my pdf into their own documentation (word). Æ becomes A, Ø becomes O and Å becomes A when copy paste from pdf in adobe reader and pasted into word. When pasting as "text only" they are fine. 

I don't expect asciidoctor-pdf to fix this, but I am wondering if you have any idea what may be the cause? This copy-paste works fine with other pdf documents, not from asciidoctor-pdf.

**Comment by mojavelinux** [comment]
> Æ becomes A, Ø becomes O and Å becomes A when copy paste from pdf in adobe reader and pasted into word.

I'm afraid this is beyond the scope of Asciidoctor PDF. It seems to be some sort of translation being applied by the operation system. I don't know the solution. What I do know is that copy-paste from PDF is not reliable in general.

**Comment by mojavelinux** [comment]
> When the upward arrow glyph is passed "on its own" - no problem. Using alpha 11.

I tried alpha 12 with your theme (using the Calibri font) and it worked fine. Keep in mind that the font must have this character.

**Comment by mojavelinux** [comment]
Original issue resolved by c3955c5cd7010dc9acb0b54f0e477ec2c81750c6.

Please file separate issues to deal with the other topics that surfaced in this thread.
